### DISCUSSION TITLE
⚡ Code faster and better - Slash Commands and Context Variables now available in Copilot via Visual Studio

### DISCUSSION THREAD

**Discussion by Akash1134** [community/community#102847] [Copilot News and Announcements]
We have introduced two new powerful features to the latest Visual Studio Copilot Chat extension: **slash commands (/)** that allow you to direct Copilot to perform specific tasks, and **context Variables (#)** that allow you to specify a file for Copilot to focus its answer on. These features are available on the latest [Chat extension](https://marketplace.visualstudio.com/items?itemName=VisualStudioExptTeam.VSGitHubCopilot) that is compatible with [Visual Studio 2022 version 17.8 and higher](https://visualstudio.microsoft.com/downloads/).

## Slash Commands[](#slash-commands) 🤖 

Slash commands are special commands that you can use in chat to receive targeted assistance, including explanations, documentation, test creation, and various other forms of support related to your code. For example, you can use:

*   /doc to add a documentation comment
*   /explain to explain the code
*   /fix to propose a fix for the problems in the selected code
*   /generate to generate code to answer your question
*   /help to get help with Copilot Chat
*   /optimize to analyze and improve the running time of the selected code
*   /tests to create unit tests for the selected code

![screenshot of slash commands](https://github.blog/wp-content/uploads/2024/01/image.jpg?w=1024&resize=1024%2C459)

## Context Variables[](#context-variables) 🎛️ 

The context variables feature enables you to add files from your solution into your questions using the # symbol. By referencing a file in this way, you allow Copilot to access the content of the file and provide more targeted answers about it. For example, you can ask “How does the #file:’Main.cs’ file work?” or “What is the purpose of the #file:’Calculator.cs’ file?” and get relevant answers from Copilot Chat. You can add multiple files to one question.

![](https://github.blog/wp-content/uploads/2024/01/image1_bcdd08.jpg?w=934&resize=934%2C1024)

That's not it! Besides these main features, you can also explore our other exciting new [preview features](https://devblogs.microsoft.com/visualstudio/copilot-chat-slash-commands-and-context-variables/) 👀 

For more such updates, keep an eye on the [GitHub blog](https://github.blog/) or better "[subscribe to our newsletter](https://github.blog/#:~:text=Subscribe%20to%20our%20newsletter)". 🚀

**Comment by matthew-es** [comment]
Akash, does the  # command for files work in VS Code as well (or is it on the list)? Thanks.

**Reply by soumya03Surya** [reply]
@thespainreport yes it works on VS Code as well. Ex: @workspace review my code as defined in #file: filename

**Reply by matthew-es** [reply]
Thanks, Soumya, but not working for me. I think everything is up-to-date. Returns:

"I'm sorry for any confusion, but as an AI, I don't have the ability to access or review specific files directly. However, I can help you improve your Python code. If you provide a code snippet or describe the issue you're facing, I'd be happy to assist you."

**Reply by sbond777** [reply]
@soumya03Surya 
@ workspace works, but #file: is **not** working for me either. Was this feature removed?

**Reply by hypebeans** [reply]
I've faced the same problem. It does work, but #file is not working either 😿

**Reply by soumya03Surya** [reply]
Seems like @workspace agent is not able to access the file referenced at times. Were you able to access the file w/o specifying workspace agent in the context  ? More here for reference -> https://code.visualstudio.com/docs/copilot/prompt-crafting#_use-context-variables

**Reply by ronnyalex** [reply]
I had problems with #file not accessing content, but I updated the extension to pre-release and #file now works. 
Before I got "Without the actual content of the....."

**Reply by ronnyalex** [reply]
Well, joy is fleeting! After coming back to work, it is not getting the content again 🤔

**Reply by ronnyalex** [reply]
Ok, so I started a New Chat and now it is working again. Maybe it triggered something
<img width="312" alt="image" src="https://github.com/community/community/assets/12079180/f3b0ea94-28bd-4901-a506-6191213a4d3c">

**Reply by popjam** [reply]
Also getting the 'this file likely contains..' and similar verbiage despite using all the necessary files as references, tried all the above suggestions but no luck. This is using @workspace and #file
